src/app/gps/qgsgpstoolbar.cpp

```diff
@@ -0,0 +1,255 @@
+        mRecenterAction->setEnabled( false );
+        mAddFeatureAction->setEnabled( false );
+        mAddTrackPointAction->setEnabled( false );
+        break;
+      case Qgis::GpsConnectionStatus::Connected:
+        whileBlocking( mConnectAction )->setChecked( true );
+        mConnectAction->setText( tr( "Disconnect GPS" ) );
+        mConnectAction->setToolTip( tr( "Disconnect from GPS" ) );
+        mConnectAction->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/gpsicons/mIconGpsDisconnect.svg" ) ) );
+        mConnectAction->setEnabled( true );
+        mRecenterAction->setEnabled( true );
+        mAddFeatureAction->setEnabled( true );
+        mAddTrackPointAction->setEnabled( mEnableAddVertexButton );
+        break;
+    }
+  } );
+
+  connect( QgisApp::instance(), &QgisApp::activeLayerChanged,
+           this, &QgsGpsToolBar::updateCloseFeatureButton );
+}
+
+void QgsGpsToolBar::setAddVertexButtonEnabled( bool enabled )
+{
+  mEnableAddVertexButton = enabled;
+  mAddTrackPointAction->setEnabled( mEnableAddVertexButton && mConnection->isConnected() );
+}
+
+void QgsGpsToolBar::updateLocationLabel( const QgsPoint &point )
+{
+  if ( point.isEmpty() )
+  {
+    mLocationLabel->clear();
+  }
+  else
+  {
+    const QString pos = QgsCoordinateUtils::formatCoordinateForProject( QgsProject::instance(), point, QgsCoordinateReferenceSystem(), 8 );
+    mLocationLabel->setText( pos );
+  }
+}
+
+void QgsGpsToolBar::updateCloseFeatureButton( QgsMapLayer *lyr )
+{
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( lyr );
+
+  if ( !( vlayer && vlayer->isValid() ) )
+    return;
+
+  // Add feature button tracks edit state of layer
+  if ( vlayer != mLastLayer )
+  {
+    if ( mLastLayer )  // disconnect previous layer
+    {
+      disconnect( mLastLayer, &QgsVectorLayer::editingStarted,
+                  this, &QgsGpsToolBar::layerEditStateChanged );
+      disconnect( mLastLayer, &QgsVectorLayer::editingStopped,
+                  this, &QgsGpsToolBar::layerEditStateChanged );
+    }
+    if ( vlayer ) // connect new layer
+    {
+      connect( vlayer, &QgsVectorLayer::editingStarted,
+               this, &QgsGpsToolBar::layerEditStateChanged );
+      connect( vlayer, &QgsVectorLayer::editingStopped,
+               this, &QgsGpsToolBar::layerEditStateChanged );
+    }
+    mLastLayer = vlayer;
+  }
+
+  QString buttonLabel = tr( "Create Feature" );
+  QString icon = QStringLiteral( "mActionCaptureLine.svg" );;
+  if ( vlayer )
+  {
+    QgsVectorDataProvider *provider = vlayer->dataProvider();
+    const QgsWkbTypes::GeometryType layerGeometryType = vlayer->geometryType();
+
+    bool enable = provider->capabilities() & QgsVectorDataProvider::AddFeatures &&  // layer can add features
+                  vlayer->isEditable() && vlayer->isSpatial();
+
+    switch ( layerGeometryType )
+    {
+      case QgsWkbTypes::PointGeometry:
+        buttonLabel = tr( "Create Point Feature at Location" );
```

Comment:
If at all appropriate / useful, consider having two modes for {single,multi}point layers.

Reply:
That's out of scope for my current project, sorry!